python/ray/data/datasource/file_based_datasource.py

```diff
@@ -467,6 +507,55 @@ def read_files(
         return read_tasks
 
 
+def _add_partitions(
+    data: Union["pyarrow.Table", "pd.DataFrame"], partitions: Dict[str, Any]
+) -> Union["pyarrow.Table", "pd.DataFrame"]:
+    import pandas as pd
+    import pyarrow as pa
+
+    assert isinstance(data, (pa.Table, pd.DataFrame))
+    if isinstance(data, pa.Table):
+        return _add_partitions_to_table(data, partitions)
+    if isinstance(data, pd.DataFrame):
+        return _add_partitions_to_dataframe(data, partitions)
+
+
+def _add_partitions_to_table(
+    table: "pyarrow.Table", partitions: Dict[str, Any]
+) -> "pyarrow.Table":
+    for field in table.column_names:
+        if field in partitions:
+            raise RuntimeError(
+                f"{field} is a partition key, but it's also the name of a column in the"
+                "read dataset."
+            )
+
+    num_columns = table.num_columns
+    for i, (field, value) in enumerate(partitions.items()):
+        column = [[value] * len(table)]
+        table = table.add_column(num_columns + i, field, column)
```

Comment:
We can use `table.append_table(field, column)` rather than having to track the index. Combining this with partition value + in-data values validation for partition keys that exist in-data (as alluded to above), this could look something like this:
```suggestion
    column_names = set(table.column_names)
    for field, value in partitions.items():
        if field in column_names:
            # Partition column exists in-data, validate that the in-data values are equal to the
            # partition value.
            import pyarrow.compute as pa
            
            # Cast partition value to column type.
            # TODO: Handle cast error.
            value = pa.scalar(value).cast(table.schema.field(field).type)
            # pac.equal() will return None if there are any null elements in the column,
            # so we explicitly check if the result is False instead of checking if its falsey.
            if pac.equal(value, table[field]) is False:
                raise ValueError(
                    f"Partition column {field} exists in table data, but partition value {value} "
                    f"is different from in-data values: {table[field].unique()}"
                )
        else:
            column = pa.array([value] * len(table))
            table = table.append_column(field, column)
```

Reply:
Addressed. `pc.equal` returned an Arrow array when I tried it, so I used `pc.all(pc.equal(...))`